### ISSUE TITLE
Several Doubts

### ISSUE THREAD

**Issue by rshaojimmy** [Jeff-sjtu/HybrIK#33]
Hi, thanks for your great work!

May I ask several questions as follows:

1. Since MSCOCO does not provide ground-truth 3D joints annotation, seems that the codes just use 2D keypoints annotation for the calculation of Joints loss?

2. In criterion.py, the implementation uses labels['target_uvd_29'] for the joints loss. This should be the regression of 3D heatmap instead of 3D joints (since 3D joints are represented by pred_xyz_jts_29)? But this paper said it does 3D keypoints estimation.

Thanks.

**Comment by Jeff-sjtu** [comment]
Hi @rshaojimmy,

For the first question: Yes, we only use the 2D keypoints for loss calculation by supervising the `uv` coordinates.
For the second question: Sorry for the misleading in the paper. We actually supervise the `uvd` coordinates for training.

**Comment by rshaojimmy** [comment]
Hi,

Thanks for your reply!

Moreover, we have run your implementation and can only achieve 50.29 PA-MPJPE (instead of 49.0 and 48.8 reported in the paper) in the 3DPW dataset. May I ask how to further improve this value to get close to the reported one?

Thanks.

**Comment by Jeff-sjtu** [comment]
Hi, may I ask how many GPUs are used for training?

**Comment by rshaojimmy** [comment]
Hi,

We just use the pre-trained hybrik model for the evaluation and we can just achieve the 50.29 PA-MPJPE in the 3DPW dataset based on your released model.

Thanks.

**Comment by Jeff-sjtu** [comment]
We have fix the bugs. I think the code works now ;)

**Comment by rshaojimmy** [comment]
yeah, it works and is a little bit better than the reported ones ;)

I notice that you use 29 joints (24 normal joints + 5 vertices based on the pre-selected leaf_number= [411, 2445, 5905, 3216, 6617]) to form the rest joints for the batch_inverse_kinematics_transform.
Actually joints and vertices are with different properties but you combine them in a single vector of rest joints.

In the inference time, you use the normal 24 joints for the evaluation.

May I ask why do you do this in the code? thanks a lot.

**Comment by Jeff-sjtu** [comment]
In the original representation of SMPL model, 24 rotations are used to represent the body pose. The `i`th rotation matrix represents the rotation of the child joint of the `i`th rotation. For the leaf joints in the original SMPL skeleton (thumbs, head, ankles), their corresponding rotations control the distal vertices. Therefore, when we conduct the inverse kinematics process, we need the coordinates of the distal vertices to solve the rotations of the leaf joints.

**Comment by rshaojimmy** [comment]
I see, thanks a lot!

**Comment by OliverSTH** [comment]
> Hi, may I ask how many GPUs are used for training?
Hi @Jeff-sjtu ,
Can I ask you if this project can use a single GPU for training or evaluation?

If so, are there any changes that need to be made?
thank you
